pkg/node/agent.go

```diff
@@ -0,0 +1,215 @@
+package node
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/coreos/bootkube/pkg/atomic"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api/v1"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+)
+
+// Agent reacts to annotation updates on the node it is running on.
+// When the annotation regarding the desired config is updated,
+// the Agent will update the configuration on disk and restart the
+// on-host kubelet.
+type Agent struct {
+	// NodeName is the name of the node the agent is running on.
+	NodeName string
+	// Client is an API server client.
+	Client clientset.Interface
+	// SysdConn is a sytemd dbus connection, used for restarting the on-host
+	// kubelet via systemd.
+	SysdConn dbusConn
+}
+
+type dbusConn interface {
+	Reload() error
+	RestartUnit(string, string, chan<- string) (int, error)
+}
+
+const (
+	// DesiredConfigAnnotation is the annotation that describes the desired
+	// configuration state for the on-host kubelet.
+	DesiredConfigAnnotation = "node-agent.alpha.coreos.com/desired-config"
+	// CurrentConfigAnnotation is the annotation that describes the current
+	// configuration state for the on-host kubelet.
+	CurrentConfigAnnotation = "node-agent.alpha.coreos.com/current-config"
+	// KubeletVersionKey is the key in the JSON stored in the config annotations that
+	// describes the version of the kubelet to run on the host.
+	KubeletVersionKey = "KUBELET_VERSION"
+	// KubeletConfigKey is the key in the JSON stored in the config annotations that
+	// describes the configuration flags for the kubelet to run on the host.
+	KubeletConfigKey = "KUBELET_CONFIG"
+
+	defaultKubeletEnvPath = "/etc/kubernetes/kubelet.env"
+	kubeletService        = "kubelet.service"
+)
+
+// NodeUpdateCallback is called via an informer watching Node objects.
+// When called, it will check the Node annotations looking for a new desired
+// config. When it finds a new desired config, it will update the on-disk config
+// and restart the on-host kubelet.
+func (a *Agent) NodeUpdateCallback(_, newObj interface{}) {
+	glog.Infof("begin node update callback")
+
+	node, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("received unexpected type: %T (expected *v1.Node)", newObj)
+		return
+	}
+
+	// Check annotations on Node.
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	var desiredConfig map[string]string
+	onDiskConfig, err := parseKubeletEnvFile(defaultKubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		goto update
+	}
+
+	// Obtain and validate the desired configuration.
+	desiredConfig, err = a.getDesiredConfig(node)
+	if err != nil {
+		glog.Error(err)
+	}
+
+update:
+	updatedNode, err := a.handleConfigUpdate(*node, onDiskConfig, desiredConfig, defaultKubeletEnvPath)
```

Comment:
`desiredConfig` could be nil here, and looks like this could cause nil pointer error in `validateConfig`